### PR TITLE
[DEV-2166] Implement API Object for UseCase

### PR DESCRIPTION
**Pull request by yanhuisg**
## Description

This PR is Part 2 of the Use Case component. It implements the API Object for UseCase. 

Note: The API Object of Context, which is one of the UseCase requirements, is not ready yet. It will be implemented in the subsequent PR

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/api/use_case.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        ),
+        ForeignKeyMapping(
+            "default_eda_table_id", ObservationTable, "default_eda_table_name", "name"
+        ),
+    ]
+
+    # pydantic instance variable (public)
+    target: Target
+    # Context api object is not implemented yet. This is a placeholder.
+    context: Optional[Any]
```
Comment (yanhuisg) [inline featurebyte/api/use_case.py:60]: Context API object is not ready yet. This is made optional temporarily until it is ready.

File: featurebyte/models/use_case.py [hunk 2]
```diff
@@ -30,7 +30,7 @@ class UseCaseModel(FeatureByteCatalogBaseDocumentModel):
         The default eda table id of from observation_table_ids
     """
 
-    context_id: PydanticObjectId
+    context_id: Optional[PydanticObjectId]
```
Comment (yanhuisg) [inline featurebyte/models/use_case.py:33]: Context API object is not ready yet. This is made optional temporarily until it is ready.

File: tests/unit/routes/test_use_case.py [hunk 3]
```diff
@@ -339,3 +339,21 @@ async def test_delete_associated_observation_table(
             f"Cannot delete Observation Table {str(new_ob_table_id)} because it is referenced"
             in response.json()["detail"]
         )
+
+    @pytest.mark.asyncio
+    async def test_delete_use_case(self, test_api_client_persistent, create_success_response):
+        """Test delete observation_table (fail) that is already associated with a use case"""
```
Comment (jevonyeoh) [inline tests/unit/routes/test_use_case.py:345]: does this comment mean that we're checking whether delete observation_table works? i might be reading it wrongly, but it looks like the test seems to just be testing whether we're deleting a use case 🤔

**Review by jevonyeoh — APPROVED** [review]
left a couple small comments

File: tests/unit/routes/test_use_case.py (hunk 3 above)
Comment (yanhuisg) [reply tests/unit/routes/test_use_case.py:345]: Yes. Because the UseCase api object also extends DeletableApiObject, so it needs to have a delete endpoint

**Pull request merged**